
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Overview</title>
    
  <link href="_static/css/theme.css" rel="stylesheet" />
  <link href="_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.e8e5499552300ddf5d7adccae7cc3b70.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/css/custom.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/js/custom.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="tutorials/installation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="#">
      
      <img src="_static/main_logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1 current">
  <a class="reference internal" href="#">
   Overview
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Tutorials
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorials/installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="tutorials/essentials/0-modelling.html">
   Essentials
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="tutorials/essentials/1-variables.html">
     Variables
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="tutorials/essentials/2-inference.html">
     Inference
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="tutorials/essentials/3-interpretation.html">
     Interpretation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="tutorials/essentials/4-model_selection.html">
     Model selection
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="tutorials/essentials/5-amortization.html">
     Amortization
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  User guide
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="guide/variables/index.html">
   Variables
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/variables/leaf_variables.html">
     Leaf variables
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/variables/custom_variables.html">
     Custom variables
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="guide/distributions/distributions.html">
   Distributions
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/distributions/mixture.html">
     Mixture distributions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/distributions/discrete.html">
     Discrete distributions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/distributions/count.html">
     Distributions
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="guide/dynamics/dynamics.html">
   Dynamics
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/dynamics/sde.html">
     Stochastic differential equations
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="guide/model_selection/index.html">
   Model selection
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="guide/model_selection/toy.html">
     Model selection
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="reference/overview.html">
   Reference
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/index.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/probabilistic-cell/docs"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/probabilistic-cell/docs/issues/new?title=Issue%20on%20page%20%2Findex.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="container">
<div class="row full-width">
    <div class="col-md-3 col-sm-4 col-lg">
        <div class="card h-100">
            <div class="card-body">
                <h3>👩‍💻 Installation</h3>
                <a href="installation.html" class="stretched-link" ></a>
                </p>
                <a class="btn btn-outline-secondary" role="button" href="#"style="z-index: 2;position: relative;">conda</a>
                <a class="btn btn-outline-secondary" role="button" href="#"style="z-index: 2;position: relative;">pip</a>
                <br>
                <code>pip install latenta lacell</code>
            </div>
        </div>
    </div>
    <div class="col-md-3 col-sm-4 col-lg">
        <div class="card h-100">
            <div class="card-body">
                <h3>🎬 The essentials</h3>
                An introduction at modelling with latenta, with a focus for (computation) biologists
                <a href="tutorials/essentials/0-modelling.html" class="stretched-link"></a>
            </div>
        </div>
    </div>
    <div class="col-md-3 col-sm-4 col-lg">
        <div class="card h-100">
            <div class="card-body">
                <h3>👨‍🎓 Tutorials</h3>
                Learning latenta by example
                <!-- <a class="btn btn-outline-secondary" role="button"></a></a>
                <a class="btn btn-outline-secondary" role="button" href="#"style="z-index: 2;position: relative;">pip</a> -->
                <a href="tutorials/overview.html" class="stretched-link"></a>
            </div>
        </div>
    </div>
</div>
<br>
<div class="row full-width">
    <div class="col-md-3 col-sm-4 col-lg">
        <div class="card h-100">
            <div class="card-body">
                <h3>👩‍🍳 User guide</h3>
                How do I do «…»?
                <a href="guide/overview.html" class="stretched-link"></a>
            </div>
        </div>
    </div>
    <div class="col-md-3 col-sm-4 col-lg">
        <div class="card h-100">
            <div class="card-body">
                <h3>🕵️‍♀️ Explanation</h3>
                Intro into the maths, statistics and machine learning
                <a href="explanation/overview.html" class="stretched-link"></a>
            </div>
        </div>
    </div>
    <div class="col-md-3 col-sm-4 col-lg">
        <div class="card h-100">
            <div class="card-body">
                <h3>🦸 Reference</h3>
                How does «…» work?
                <a href="reference/overview.html" class="stretched-link"></a>
            </div>
        </div>
    </div>
</div>
</div>
<br>
<div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>Latenta models the cell as a combination of different <em>cellular processes</em> (cell types, cell cycle, zonation, genetic perturbations, time, unexplained variation, nuissance variables, …). Each of these processes influence <em>observations</em> (transcriptome, unspliced transcriptome, surface proteome, chromatin, phenotype, …). The processes may also interact with each other, and can thus combinatorially affect observations in various ways. Latenta is modular by design, and this is particularly useful when working with multimodal data, data where cells have multiple co-variates (including nuissance variables), or when trying to find interpretable <em>mechanistic</em> models. In this way, it allows you to answer questions such as:</p>
<ul class="simple">
<li><p>How does the remaining variation look like once I model («regress out») my time series, batch effect and cell cycle?</p></li>
<li><p>Does my genetic perturbation affect target gene expression in a linear, sigmoid, exponential or more complex manner?</p></li>
<li><p>How is the cell cycle distributed with zonation?</p></li>
<li><p>Are there genes that are induced after 6 hours commonly in all celltypes?</p></li>
<li><p>Does the surgery «interact» with the cell cycle?</p></li>
<li><p>Do my genetic perturbations interact? Is this in an additive, synergistic, or more complex <em>epistatic</em> manner?</p></li>
<li><p>How many cells are cycling in a spatial spot, and in which phases are these cells?</p></li>
<li><p>Is my genetic perturbation inducing two or three cellular states?</p></li>
</ul>
<p>Key concepts are:</p>
<ul class="simple">
<li><p><strong>Modular</strong>: Although each biological dataset and question is unique, different processes come up again and again and can thus be reused across models. Latent contains different standard modules to different cellular processes, such as a clustering, a batch effect, a transcriptome, a linear trajectory, … and these can be combined in ways to suit a particular dataset.</p></li>
<li><p><strong>Flexible</strong>: Technology and biology is still evolving, and we cannot predict what kind of models will be necessary to model these new data types. Moreover, most technical parameters, such as a transcriptome’s library size, can also have a biological meaning. Latenta provides the flexibility to adapt a model to a particular use case, and to extend</p></li>
<li><p><strong>Scalable</strong>: Single-cell omics datasets are relatively large, both in number of cells and in number of features. Moreover, a user will likely require several modelling steps and iterations ideally done in an interactive fashion.</p></li>
<li><p><strong>Statistical significance</strong>: Despite large cell numbers, the actual number of biological replicates (tissues, patients, …), is relatively low. Latenta therefore contains tools to assess statistical significance of your findings.</p></li>
<li><p><strong>Model selection</strong>: We typically don’t know the underlying biochemical mechanism that drives certain biological processes. In fact, understanding this mechanism is often one desired outcome of the study. Latenta thus contains methods to do model selection, and to check whether some models are more or equally likely than others.</p></li>
<li><p><strong>Hidden latent variables</strong>: Interesting biological phenomena, such as cell differentiation or the cell cycle, are often unobserved. Latenta allows you to model these hidden latent variables and relate them to observation.</p></li>
<li><p><strong>Hypothetical latent variables</strong>: If we want to do an unbiased analysis, or want to model any unexplained variation, we can put these into hypothetical latent variables (such as a clustering or an embedding). Latenta contains tools to model these variables and combine them with hidden latent variables (i.e. “regressing out”).</p></li>
<li><p><strong>Interpretability</strong>: Interpretable models are often preferable, both for the latent space (for example, latent variables that denote the cell cycle) and any functions that connect the latent space to observations (for example, a hill equation). Latenta’s models therefore often focus on interpretability, and only use black-box functions such as neural networks when extreme flexibility is required.</p></li>
<li><p><strong>Prior information</strong>: A lot is already known about certain biological processes that we’re studying, and it is desirable to include this knowledge so that we can focus on what is unknown. Latenta therefore uses Bayesian modelling to include prior knowledge wherever available.</p></li>
<li><p><strong>Hierarchical interactions</strong>: Biological processes are tightly integrated and interconnected. For example, a differentiation process almost always affects the activity of the cell cycle. Latenta can therefore not only connect the latent space to observations, but also create connections between latent variables themselves.</p></li>
<li><p><strong>Minimal analytical mindset</strong>: Any framework that follows the previous principles will likely require considerable mathematical and statistical know-how to implement.</p></li>
</ul>
<p>For example, a cell cycle model can be fitted once and then reused across single-cell and spatial datasets. Or, a latent variable can be fit on one modality (e.g. the transcriptome) and then used to model other modalities. But most importantly, having one unified framework allows the simultaneous modelling different processes, whether you are modelling scalar variables (trajectories, zonation, time), circular variables (cell cycle, circadian clock), discrete variables (clustering, cell type, cell state, batch effect), vector variables (embedding, proportions), nuissance variables (batch, ambient mRNA), different modalities (transcriptome, surface proteome) or different ways of sampling (spatial, nuclear, RNA-FISH, …).</p>
<p>Some variables may be observed (cell types, transcriptome, …), while other <em>latent</em> variables may need to be inferred (cell cycle, cell state, …). Latenta provides tools to model these processes properly, include prior information where needed, do biological and statistical interpretation, and transfer models between datasets.</p>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='right-next' id="next-link" href="tutorials/installation.html" title="next page">Installation</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By the latenta team<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>